openmls/src/ciphersuite/crypto/rust_crypto_provider.rs

```diff
@@ -0,0 +1,214 @@
+                .map(|r| r.as_slice().into())
+                .map_err(|_| CryptoError::AeadDecryptionError)
+        }
+        AeadType::ChaCha20Poly1305 => {
+            let aes = ChaCha20Poly1305::new_from_slice(key)
+                .map_err(|_| CryptoError::CryptoLibraryError)?;
+            aes.encrypt(nonce.into(), Payload { msg: ct_tag, aad })
+                .map(|r| r.as_slice().into())
+                .map_err(|_| CryptoError::AeadDecryptionError)
+        }
+        _ => Err(CryptoError::UnsupportedAeadAlgorithm),
+    }
+}
+
+/// Returns `(sk, pk)`
+pub(crate) fn signature_key_gen(alg: SignatureScheme) -> Result<(Vec<u8>, Vec<u8>), CryptoError> {
+    match alg {
+        SignatureScheme::ECDSA_SECP256R1_SHA256 => {
+            let k = SigningKey::random(&mut OsRng);
+            let pk = k.verifying_key().to_encoded_point(false).as_bytes().into();
+            Ok((k.to_bytes().as_slice().into(), pk))
+        }
+        SignatureScheme::ED25519 => {
+            let k = ed25519_dalek::Keypair::generate(&mut rand_07::rngs::OsRng).to_bytes();
+            let pk = k[32..].to_vec();
+            // full key here because we need it to sign...
+            let sk_pk = k.into();
+            Ok((sk_pk, pk))
+        }
+        _ => Err(CryptoError::UnsupportedSignatureScheme),
+    }
+}
+
+pub(crate) fn verify_signature(
+    alg: SignatureScheme,
+    data: &[u8],
+    pk: &[u8],
+    signature: &[u8],
+) -> Result<(), CryptoError> {
+    match alg {
+        SignatureScheme::ECDSA_SECP256R1_SHA256 => {
+            let k = VerifyingKey::from_encoded_point(
+                &EncodedPoint::from_bytes(pk).map_err(|_| CryptoError::CryptoLibraryError)?,
+            )
+            .map_err(|_| CryptoError::CryptoLibraryError)?;
+            k.verify(
+                data,
+                &Signature::from_der(signature).map_err(|_| CryptoError::InvalidSignature)?,
+            )
+            .map_err(|_| CryptoError::InvalidSignature)
+        }
+        SignatureScheme::ED25519 => {
+            let k = ed25519_dalek::PublicKey::from_bytes(pk)
+                .map_err(|_| CryptoError::CryptoLibraryError)?;
+            if signature.len() != 64 {
+                return Err(CryptoError::CryptoLibraryError);
+            }
+            let mut sig = [0u8; 64];
+            sig.clone_from_slice(signature);
+            k.verify_strict(data, &ed25519_dalek::Signature::new(sig))
+                .map_err(|_| CryptoError::InvalidSignature)
+        }
+        _ => Err(CryptoError::UnsupportedSignatureScheme),
+    }
+}
+
+pub(crate) fn sign(alg: SignatureScheme, data: &[u8], key: &[u8]) -> Result<Vec<u8>, CryptoError> {
+    match alg {
+        SignatureScheme::ECDSA_SECP256R1_SHA256 => {
+            let k = SigningKey::from_bytes(key).map_err(|_| CryptoError::CryptoLibraryError)?;
+            let signature = k.sign(data);
+            Ok(signature.to_der().to_bytes().into())
```

Comment:
Does this make #452 specific to evercrypt?

Reply:
Yes, the code from  #452 will move into the `evercrypt_provider.rs`.